src/coreclr/src/vm/eventpipesession.cpp

```diff
@@ -482,12 +537,32 @@ void EventPipeSession::SuspendWriteEvent()
         THROWS;
         GC_TRIGGERS;
         MODE_PREEMPTIVE;
+        // Need to disable the session before calling this method
+        PRECONDITION(!EventPipe::IsSessionEnabled(reinterpret_cast<EventPipeSessionID>(this)));
     }
     CONTRACTL_END;
 
-    // Force all in-progress writes to either finish or cancel
-    // This is required to ensure we can safely flush and delete the buffers
-    m_pBufferManager->SuspendWriteEvent(GetIndex());
+    {
+        SpinLockHolder holder(EventPipeThread::GetGlobalThreadLock());
```

Comment:
I meant to comment and forgot, we shouldn't spin waiting for writes to complete inside this lock. WriteEvent() isn't guaranteed to be fast in all cases and all thread creation/destruction in the process has to block on this lock. I'd recommend copying the set of currently active threads into a new snapshot of the list, then iterate over the snapshot outside the lock to wait for in progress writes to complete.